### DISCUSSION TITLE
Soft Body Physics

### DISCUSSION THREAD

**Discussion by neverwinter-nights** [o3de/o3de#15859] [Help Wanted]
Hello.

Does O3DE support soft body physics by nvidia PhysX ?

https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/SoftBodies.html
https://www.o3de.org/docs/user-guide/interactivity/physics/nvidia-physx/simulated-bodies/

If no, are there any plans to implement it ?

Thank you.

**Comment by Nolram12345** [comment]
PhysX 5 softbodies unfortunately only work with CUDA acceleration, meaning that they are only usable on NVIDIA cards that support CUDA. I feel like a better alternative, that may actually allow for some other interesting effects in the future, may be AMD's FEMFX which runs on the CPU, so it works everywhere. https://github.com/GPUOpen-Effects/FEMFX

**Reply by anon-oss** [reply]
FEMFX only work on windows and it seems that it is abandoned by GpuOpen. There were some community efforts to make it work on linux which reached no where due to the size of the project. 
Softbody middleware would have to interact with the default physics engine too.

**Reply by Nolram12345** [reply]
Oh yeah, I didn't realise it was Windows only... I do wonder if its possible to fork though for wider platform support or implement a system based on it. Because the technology and methods behind it are immensely powerful; not only providing simple softbodies, but complex body simulation including fracturing for incredibly realistic destruction. Its basically an open-source equivalent (okay, not quite, not similar) of the old DMM middleware.

**Reply by anon-oss** [reply]
[This incomplete linux fork ](https://github.com/w23/FEMFX/tree/linux-port) might be be a good starting point for wider platform support. For implementation will have to look at gpuopen FEMFX integration plugins for various engines. We will have to impliment it first on windows before other platforms. 
But since FEMFX was abandoned by amd soon after release will it be good enough in 5 years given that other open physics engines like bullet physics has some form of basic softbody physics?

**Reply by Nolram12345** [reply]
The key difference here is performance and featureset. Yes, Bullet does have softbodies - but they are quite slow. FEMFX is directly optimised for games. The reason FEMFX got "abandoned" is because a couple of the key engineers behind it, a lot of them originally from Pixelux who made the infamous DMM physics engine (which is also internally an FEM simulator, shares a lot in common with FEMFX) left AMD a couple years after FEMFX release. 
The FEMFX simulation backend is notable for being remarkably optimised, even while being a full FEM simulator. Unlike Bullet, which has mainly occupied VFX and Robotics spaces, it was made for games and makes extensive use of SIMD.

FEMFX also has some neat features that I think could positively impact O3DE, such as fracturing (it could be used a destruction engine for simulation lots of materials with bending, such as wood, or stiffer ones too like brick). The biggest issue I would see besides the actual implementation is getting the workflow perhaps a bit nicer, but with O3DE's asset pipeline I think it may be possible to work around most of the limitations FEMFX had as an independent library that can't directly rely on engine systems.
